Implement Callback Providers for dynamic file lists

robertscherer opened the pull request:
As discussed with @markstory on IRC today, it would be nice to be able to add callback strings to the INI configuration, which modify the files[] list dynamically.

This is a first, naive implementation of how I would imagine this could work.

I'm mainly unsure about the position in the process the conversion from a callback string to a file list should happen.

I have implemented this in AssetConfig which is the soonest point to do this - which I think is a sensitive choice, but I could be wrong. 

An alternative would be moving this into the asset/file scanning part of the process, but this could have unintended side effects.

Also, I'm unsure about how to handle errors - should Exceptions be thrown if the callback isn't callable or doesn't return an array?

I'm happy to get some feedback and move this further.

Cheers,
Robert Scherer

robertschererc committed:
Prototype implementation of how callback file providers could work.

markstory commented on src/AssetConfig.php:
```diff
@@ -525,6 +528,30 @@ public function addTarget($target, array $config)
+                }
+            }
+        }
+        return $config;
```
Other expansions are currently happening in `MiniAsset\Factory::target()`. It current handles remote/local files, will soon handle glob expansion. I see this as similar to glob expansion so perhaps it belongs there?

markstory commented:
Other than the placement this looks good. There are a few [PHPCS errors](https://travis-ci.org/markstory/mini-asset/jobs/59583393) that I can clean up afterwards if you want.

robertschererc committed:
Move callback provider functionality from AssetConfig to Factory

robertscherer commented:
Ok - I have moved the functionality to `Factory` and fixed the cs errors.

Preparing the files list outside of the loop has the advantage that callback providers could return a mix of `Local` and `Remote` files. Looking at the glob implementation, the callback providers could return glob patterns and it would work fine.

robertschererc committed:
- Throw a RuntimeException if a callback provider is not callable
- Additional Unit Test for checking if file ordering when mixing local with callback provided files

markstory commented:
Thanks :+1: I'll update the documentation.
